src/core/meshcop/tcat_agent.cpp

```diff
@@ -629,13 +656,135 @@ Error TcatAgent::HandleGetProvisioningUrl(Message &aOutgoingMessage, bool &respo
+                       StringLength(mVendorInfo->mInstallCode, 255));
+
+exit:
+    return error;
+}
+
+Error TcatAgent::HandleRequestRandomNumberChallenge(Message &aOutgoingMessage, bool &response)
+{
+    Error error = kErrorNone;
+
+    SuccessOrExit(error = Random::Crypto::Fill(mRandomChallenge));
+
+    SuccessOrExit(
+        error = Tlv::AppendTlv(aOutgoingMessage, kTlvResponseWithPayload, &mRandomChallenge, sizeof(mRandomChallenge)));
+    response = true;
+exit:
+    return error;
+}
+
+Error TcatAgent::HandleRequestPskdHash(const Message &aIncomingMessage,
+                                       Message       &aOutgoingMessage,
+                                       uint16_t       aOffset,
+                                       uint16_t       aLength,
+                                       bool          &response)
+{
+    Error                    error             = kErrorNone;
+    uint64_t                 providedChallenge = 0;
+    Crypto::HmacSha256::Hash hash;
+
+    VerifyOrExit(StringLength(mVendorInfo->mPskdString, kMaxPSKdLength) != 0, error = kErrorFailed);
+    VerifyOrExit(aLength == sizeof(providedChallenge), error = kErrorParse);
+
+    SuccessOrExit(error = aIncomingMessage.Read(aOffset, &providedChallenge, aLength));
+
+    hash = CalculateHash(providedChallenge, mVendorInfo->mPskdString,
+                         StringLength(mVendorInfo->mPskdString, kMaxPSKdLength));
+
+    SuccessOrExit(error = Tlv::AppendTlv(aOutgoingMessage, kTlvResponseWithPayload, hash.GetBytes(),
+                                         Crypto::HmacSha256::Hash::kSize));
     response = true;
 
 exit:
     return error;
 }
 
+Error TcatAgent::VerifyHash(const Message &aIncomingMessage,
+                            uint16_t       aOffset,
+                            uint16_t       aLength,
+                            const char    *aBuf,
+                            size_t         aBufLen)
+{
+    Error                    error = kErrorNone;
+    Crypto::HmacSha256::Hash hash;
+    Crypto::HmacSha256::Hash recivedHash;
+
+    VerifyOrExit(aLength == Crypto::HmacSha256::Hash::kSize, error = kErrorSecurity);
+    VerifyOrExit(mRandomChallenge != 0, error = kErrorSecurity);
+
+    SuccessOrExit(error = aIncomingMessage.Read(aOffset, recivedHash.m8, aLength));
+
+    hash = CalculateHash(mRandomChallenge, aBuf, aBufLen);
+
+    VerifyOrExit(hash == recivedHash, error = kErrorSecurity);
+
+exit:
+    return error;
+}
+
+Crypto::HmacSha256::Hash TcatAgent::CalculateHash(uint64_t aChallenge, const char *aBuf, size_t aBufLen)
+{
+    mbedtls_asn1_buf        *rawKey = Get<Ble::BleSecure>().GetOwnPublicKey();
+    Crypto::Key              cryptoKey;
+    Crypto::HmacSha256       hmac;
+    Crypto::HmacSha256::Hash hash;
+
+    cryptoKey.Set(rawKey->p, static_cast<uint16_t>(rawKey->len));
+
+    hmac.Start(cryptoKey);
+    hmac.Update(aChallenge);
+    hmac.Update(aBuf, static_cast<uint16_t>(aBufLen));
+    hmac.Finish(hash);
+
+    return hash;
```

Comment:
This works fine, but some embedded toolchains do extra copying when we return more complex types from a method. 

Can consider to instead pass a reference to a `Crypto::HmacSha256::Hash` as input arg to this method for it to populate?

Reply:
fixed.